### ISSUE TITLE
[java] UnnecessaryCast  false positive with package private methods

### ISSUE THREAD

**Issue by hgschmie** [pmd/pmd#5732]
**Affects PMD Version:** 

7.13.0 

**Rule:**

Please provide the rule name and a link to the rule documentation:
<https://docs.pmd-code.org/latest/pmd_rules_java_codestyle.html#unnecessarycast>

**Description:**

PMD gets confused with package private methods.

**Code Sample demonstrating the issue:**

```java
package A;

public class Foo {
    String getAThing() {
        return "a thing";
    }
}
```

```java
package B;

public class Bar extends Foo {
    public String getBThing() {
        return "b thing";
    }
}
```

```java
package A;

import B.Bar;

public class Test {
    public static void main(String ... args) {
        Bar bar = new Bar();
        System.out.println(((Foo) bar).getAThing());
        System.out.println(bar.getBThing());
    }
}
```

The `Test` class must cast the `Bar` to a `Foo` to access the package private method `getAThing()`, otherwise it is package private in the `B` package, not the `A` package.

PMD reports `PMD Failure: A.Test:8 Rule:UnnecessaryCast Priority:3 Unnecessary cast (Foo).` for the line with the cast.

**Expected outcome:**

PMD should realize that, for a package private method, the cast is necessary to bring the object into the right package for package private access.

PMD reports a violation at line 8, but that's wrong. That's a false positive.

**Running PMD through:** *Maven*
